docs/build.md

```diff
@@ -280,6 +280,46 @@ The following compilation options are also available to tweak performance (yes,
 
 **Windows**
 
+#### Git Bash MINGW64
+
+[![](https://i.imgur.com/M8dnyk5.png)](https://git-scm.com/downloads/win).
+
+Download and install [`Git-SCM`](https://git-scm.com/downloads/win) with the default settings
+
+[![](https://i.imgur.com/hiAMq2S.png)](https://visualstudio.microsoft.com/)
+
+Download and install [`Visual Studio Community Edition`](https://visualstudio.microsoft.com/) and make sure you select `C++`
```

Comment:
I am confused by this, why is VS necessary to build with MINGW64?

Reply:
Installing visual studio installs the `Windows SDK`

Reply:
```
J@BUD MINGW64 ~/Desktop/llm/llama.cpp (master)
$ cmake -B build -DGGML_VULKAN=ON
-- Building for: Visual Studio 17 2022
-- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.22631.
-- The C compiler identification is MSVC 19.41.34120.0
-- The CXX compiler identification is MSVC 19.41.34120.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.41.34120/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.41.34120/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Git: C:/Program Files/Git/mingw64/bin/git.exe (found version "2.47.0.windows.1")
```

Reply:
> I am confused by this, why is VS necessary to build with MINGW64?

Installing `Visual Studio` installs the `Windows SDK`, I think you need it for this.

I guess you could also just install the [`Windows SDK Directly`](https://developer.microsoft.com/en-us/windows/downloads/windows-sdk/)

Reply:
> I am confused by this, why is VS necessary to build with MINGW64?

```
J@BUD MINGW64 ~/Desktop/llm/llama.cpp (master)
$ cmake -B build -DGGML_VULKAN=ON
-- Building for: Visual Studio 17 2022
-- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.22631.
-- The C compiler identification is MSVC 19.41.34120.0
-- The CXX compiler identification is MSVC 19.41.34120.0
```

Installing `Visual Studio` installs the `Windows SDK`, I think you need it for this.

I guess you could also just install the [`Windows SDK Directly`](https://developer.microsoft.com/en-us/windows/downloads/windows-sdk/)

Reply:
I already have visual studio installed so the reason I added it to the list is because I see `Selecting Windows SDK` and I know it came with visual studio and not git-scm, here is the output from the build

```
J@BUD MINGW64 ~/Desktop/llm/llama.cpp (master)
$ cmake -B build -DGGML_VULKAN=ON
-- Building for: Visual Studio 17 2022
-- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.22631.
-- The C compiler identification is MSVC 19.41.34120.0
-- The CXX compiler identification is MSVC 19.41.34120.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.41.34120/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.41.34120/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Git: C:/Program Files/Git/mingw64/bin/git.exe (found version "2.47.0.windows.1")
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Building for: Visual Studio 17 2022
-- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.22631.
-- The C compiler identification is MSVC 19.41.34120.0
-- The CXX compiler identification is MSVC 19.41.34120.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.41.34120/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.41.34120/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Git: C:/Program Files/Git/mingw64/bin/git.exe (found version "2.47.0.windows.1")
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - not found
-- Found Threads: TRUE
-- Warning: ccache not found - consider installing it for faster compilation or disable this warning with GGML_CCACHE=OFF
-- CMAKE_SYSTEM_PROCESSOR: AMD64
-- CMAKE_GENERATOR_PLATFORM:
-- Found OpenMP_C: -openmp (found version "2.0")
-- Found OpenMP_CXX: -openmp (found version "2.0")
-- Found OpenMP: TRUE (found version "2.0")
-- OpenMP found
-- Using llamafile
-- x86 detected
-- Performing Test HAS_AVX_1
-- Performing Test HAS_AVX_1 - Success
-- Performing Test HAS_AVX2_1
-- Performing Test HAS_AVX2_1 - Success
-- Performing Test HAS_FMA_1
-- Performing Test HAS_FMA_1 - Success
-- Performing Test HAS_AVX512_1
-- Performing Test HAS_AVX512_1 - Failed
-- Performing Test HAS_AVX512_2
-- Performing Test HAS_AVX512_2 - Failed
-- Using runtime weight conversion of Q4_0 to Q4_0_x_x to enable optimized GEMM/GEMV kernels
-- Including CPU backend
CMake Warning at ggml/src/ggml-amx/CMakeLists.txt:106 (message):
  AMX requires x86 and gcc version > 11.0.  Turning off GGML_AMX.


-- Found Vulkan: C:/VulkanSDK/1.3.296.0/Lib/vulkan-1.lib (found version "1.3.296") found components: glslc glslangValidator
-- Vulkan found
-- Including Vulkan backend
-- Configuring done (26.5s)
-- Generating done (0.7s)
-- Build files have been written to: C:/Users/J/Desktop/llm/llama.cpp/build
```

Reply:
I'll set up a virtual machine and see if I can skip installing visual studio

Reply:
> I am confused by this, why is VS necessary to build with MINGW64?

```
J@BUD MINGW64 ~/Desktop/llm/llama.cpp (master)
$ cmake -B build -DGGML_VULKAN=ON
-- Building for: Visual Studio 17 2022
-- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.22631.
-- The C compiler identification is MSVC 19.41.34120.0
-- The CXX compiler identification is MSVC 19.41.34120.0
```

Installing `Visual Studio` installs `MSVC` and the `Windows SDK`

I guess you could also just install the [`Windows SDK Directly`](https://developer.microsoft.com/en-us/windows/downloads/windows-sdk/) and download the [`VC Redistributables`](https://learn.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170) but I know this works.

Reply:
```
J@BUD MINGW64 ~/Desktop/llm/llama.cpp (master)
$ cmake -B build -DGGML_VULKAN=ON
-- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.22631.
-- Warning: ccache not found - consider installing it for faster compilation or disable this warning with GGML_CCACHE=OFF
-- CMAKE_SYSTEM_PROCESSOR: AMD64
-- CMAKE_GENERATOR_PLATFORM:
-- Building for: Visual Studio 17 2022
-- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.22631.
-- The C compiler identification is MSVC 19.41.34120.0
-- The CXX compiler identification is MSVC 19.41.34120.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.41.34120/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.41.34120/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Git: C:/Program Files/Git/mingw64/bin/git.exe (found version "2.47.0.windows.1")
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - not found
-- Found Threads: TRUE
-- Warning: ccache not found - consider installing it for faster compilation or disable this warning with GGML_CCACHE=OFF
-- CMAKE_SYSTEM_PROCESSOR: AMD64
-- CMAKE_GENERATOR_PLATFORM:
-- Found OpenMP_C: -openmp (found version "2.0")
-- Found OpenMP_CXX: -openmp (found version "2.0")
-- Found OpenMP: TRUE (found version "2.0")
-- OpenMP found
-- Using llamafile
-- x86 detected
-- Performing Test HAS_AVX_1
-- Performing Test HAS_AVX_1 - Success
-- Performing Test HAS_AVX2_1
-- Performing Test HAS_AVX2_1 - Success
-- Performing Test HAS_FMA_1
-- Performing Test HAS_FMA_1 - Success
-- Performing Test HAS_AVX512_1
-- Performing Test HAS_AVX512_1 - Failed
-- Performing Test HAS_AVX512_2
-- Performing Test HAS_AVX512_2 - Failed
-- Using runtime weight conversion of Q4_0 to Q4_0_x_x to enable optimized GEMM/GEMV kernels
-- Including CPU backend
CMake Warning at ggml/src/ggml-amx/CMakeLists.txt:106 (message):
  AMX requires x86 and gcc version > 11.0.  Turning off GGML_AMX.


-- Found Vulkan: C:/VulkanSDK/1.3.296.0/Lib/vulkan-1.lib (found version "1.3.296") found components: glslc glslangValidator
-- Vulkan found
-- Including Vulkan backend
-- Configuring done (26.9s)
-- Generating done (0.6s)
-- Build files have been written to: C:/Users/J/Desktop/llm/llama.cpp/build
```

If you don't think visual studio is needed you can remove the reviewed lines

Reply:
you could install the [`Windows SDK Directly`](https://developer.microsoft.com/en-us/windows/downloads/windows-sdk/) and download the [`VC Redistributables`](https://learn.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170) and the [`.NET SDK for MSBuild`](https://learn.microsoft.com/en-us/visualstudio/msbuild/whats-new-msbuild-17-0?view=vs-2022) instead of installing visual studio

Reply:
In theory you could install the [`Windows SDK Directly`](https://developer.microsoft.com/en-us/windows/downloads/windows-sdk/) and download the [`VC Redistributables`](https://learn.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170) and the [`.NET SDK for MSBuild`](https://learn.microsoft.com/en-us/visualstudio/msbuild/whats-new-msbuild-17-0?view=vs-2022) instead of installing visual studio, the goal is to keep the instructions simple.

Reply:
At this level of complexity you'd be better off with https://github.com/skeeto/w64devkit , which is also portable, unlike MSVS.

Reply:
> At this level of complexity you'd be better off with https://github.com/skeeto/w64devkit , which is also portable, unlike MSVS.

These are all things you probably already have on your computer and `w64devkit` is flagged as a virus so I can't use it.

Reply:
> I am confused by this, why is VS necessary to build with MINGW64?

```
J@BUD MINGW64 ~/Desktop/llm/llama.cpp (master)
$ cmake -B build -DGGML_VULKAN=ON
-- Building for: Visual Studio 17 2022
-- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.22631.
-- The C compiler identification is MSVC 19.41.34120.0
-- The CXX compiler identification is MSVC 19.41.34120.0
```
```
J@BUD MINGW64 ~/Desktop/llm/llama.cpp (master)
$ cmake --build build --config Release
MSBuild version 17.11.2+c078802d4 for .NET Framework

  1>Checking Build System
  Generating build details from Git
```

Installing `Visual Studio` installs  the `Windows SDK`, `MSVC` and `MSBuild`

Reply:
> At this level of complexity you'd be better off with `w64devkit` 

Its flagged as a virus, I can't use it.

This isn't complex.

Reply:
> Its flagged as a virus, I can't use it.

Which is false positive, as always (thanks, MS). You can download [previous release](https://github.com/skeeto/w64devkit/releases/tag/v1.23.0) in zip format, I use it currently without problems.

I understand that portability is a unique feature of Windows (and similar systems, like Symbian), but it's very handy and, unfortunately, overlooked.

Reply:
No it isn't.

Reply:
It doesn't change the fact that I can't use it.

Reply:
All of the steps are easy and necessary and this is a valid way to build vulkan `llama.cpp`

I don't really understand what the resistance is, "there are other ways" doesn't seem like a valid argument.

Reply:
I don't really know what that means, when I was reading the build guide I seen that vulkan needed mingw and remembered that git-scm had it and it just worked so I made this guide.

Reply:
I moved my edit below the option that you prefer.

for people who already have these things installed this is a lot easier.

Reply:
Having documentation for using MINGW64 directly is good. I feel that the images are overkill, I would prefer text-only instructions like the rest of the documentation. People compiling llama.cpp themselves should already have the basic knowledge to know how to do these things without images, otherwise they can use the pre-built binaries in the releases.

Reply:
okay, I will remove the images

Reply:
I removed the images I think this is resolved now